File: api/app/normalize.py
"""1:1 port of web/src/normalizeState.ts, operating on a raw dict before
Pydantic validation. Needed so the import endpoint can accept the same
older/legacy-format project files the frontend's local-mode import already
handles (missing fields filled with defaults, the deprecated single-nail
box-relative-percentage format migrated to the current cm-offset array) —
without this, an old export would just fail Pydantic validation instead of
being migrated. Keep in sync with normalizeState.ts if that ever changes.
"""


def normalize_state(parsed: dict) -> dict:
    if not isinstance(parsed, dict) or not parsed.get("wall") or not isinstance(parsed.get("images"), list):
        raise ValueError("Not a wall-projector project file")

    if not parsed.get("ruler"):
        parsed["ruler"] = {"length": 100, "visible": True, "color": "#ffcc00"}
    if not parsed["ruler"].get("color"):
        parsed["ruler"]["color"] = "#ffcc00"

    if not parsed.get("background"):
        parsed["background"] = {"enabled": False, "color": "#2a2a2a", "projectToo": False}
    if parsed["background"].get("projectToo") is None:
        parsed["background"]["projectToo"] = False

    if not parsed.get("defaults"):
        parsed["defaults"] = {"imageWidth": 30, "frameEnabled": False, "frameColor": "black", "frameWidth": 3}

    if not parsed.get("grid"):
        parsed["grid"] = {"enabled": False, "size": 20, "projectToo": False}
    if parsed["grid"].get("projectToo") is None:
        parsed["grid"]["projectToo"] = False

    if not parsed.get("nail"):
        parsed["nail"] = {"enabled": False, "color": "#ff3b3b", "size": 10}

    if not parsed.get("keystone"):
        parsed["keystone"] = {"enabled": False, "vertical": 0, "horizontal": 0}

    wall = parsed["wall"]
    for im in parsed["images"]:
        if not im.get("frame"):
            im["frame"] = {"enabled": False, "color": "black", "width": 3}
        if not isinstance(im.get("nails"), list):
            w_real = (im["wPct"] / 100) * wall["width"]
            h_real = (im["hPct"] / 100) * wall["height"]
            if im.get("nailXPct") is not None and im.get("nailYPct") is not None:
                # migrate from the old single-nail, box-relative-percentage format
                im["nails"] = [{"xCm": (im["nailXPct"] / 100) * w_real, "yCm": (im["nailYPct"] / 100) * h_real}]
            else:
                im["nails"] = [{"xCm": w_real / 2, "yCm": h_real / 2}]
        im.pop("nailXPct", None)
        im.pop("nailYPct", None)
        if im.get("aspectLocked") is None:
            im["aspectLocked"] = True
        if im.get("crop") is None:
            im["crop"] = False
        if im.get("snapToGrid") is None:
            im["snapToGrid"] = False

    return parsed

File: src/lib/project/normalize.ts
import type { WallProjectState } from './state';

/**
 * Fills in defaults for any fields missing from an older save (or a file
 * from an older version of the app) and migrates deprecated formats. Shared
 * by every load path (DB read normalization happens at write time via the
 * PUT route's schema, import happens here; the frontend also calls this
 * directly for localStorage loads/file imports), so all apply the exact
 * same rules. Mutates and returns the given object; throws if it doesn't
 * look like a wall-projector state at all. Ported from the pre-rebuild
 * app.js / web/src/normalizeState.ts and api/app/normalize.py — kept as the
 * one copy now that frontend and backend share a codebase.
 */
export function normalizeState(parsed: unknown): WallProjectState {
  if (
    !parsed ||
    typeof parsed !== 'object' ||
    !(parsed as any).wall ||
    !Array.isArray((parsed as any).images)
  ) {
    throw new Error('Not a wall-projector project file');
  }

  const state = parsed as any;

  if (!state.ruler) state.ruler = { length: 100, visible: true, color: '#ffcc00' };
  if (!state.ruler.color) state.ruler.color = '#ffcc00';

  if (!state.background) state.background = { enabled: false, color: '#2a2a2a', projectToo: false };
  if (state.background.projectToo === undefined) state.background.projectToo = false;

  if (!state.defaults) {
    state.defaults = { imageWidth: 30, frameEnabled: false, frameColor: 'black', frameWidth: 3 };
  }

  if (!state.grid) state.grid = { enabled: false, size: 20, projectToo: false };
  if (state.grid.projectToo === undefined) state.grid.projectToo = false;

  if (!state.nail) state.nail = { enabled: false, color: '#ff3b3b', size: 10 };

  if (!state.keystone) state.keystone = { enabled: false, vertical: 0, horizontal: 0 };

  for (const im of state.images) {
    if (!im.frame) im.frame = { enabled: false, color: 'black', width: 3 };
    if (!Array.isArray(im.nails)) {
      const wReal = (im.wPct / 100) * state.wall.width;
      const hReal = (im.hPct / 100) * state.wall.height;
      if (im.nailXPct !== undefined && im.nailYPct !== undefined) {
        // migrate from the old single-nail, box-relative-percentage format
        im.nails = [{ xCm: (im.nailXPct / 100) * wReal, yCm: (im.nailYPct / 100) * hReal }];
      } else {
        im.nails = [{ xCm: wReal / 2, yCm: hReal / 2 }];
      }
    }
    delete im.nailXPct;
    delete im.nailYPct;
    if (im.aspectLocked === undefined) im.aspectLocked = true;
    if (im.crop === undefined) im.crop = false;
    if (im.snapToGrid === undefined) im.snapToGrid = false;
  }

  return state as WallProjectState;
}
